crates/ty_python_semantic/src/types.rs

```diff
@@ -11765,33 +11771,55 @@ impl<'db> ModuleLiteralType<'db> {
                 .member(db, "__dict__");
         }
 
-        // If the file that originally imported the module has also imported a submodule
-        // named `name`, then the result is (usually) that submodule, even if the module
-        // also defines a (non-module) symbol with that name.
-        //
-        // Note that technically, either the submodule or the non-module symbol could take
-        // priority, depending on the ordering of when the submodule is loaded relative to
-        // the parent module's `__init__.py` file being evaluated. That said, we have
-        // chosen to always have the submodule take priority. (This matches pyright's
-        // current behavior, but is the opposite of mypy's current behavior.)
-        if self.available_submodule_attributes(db).contains(name) {
-            if let Some(submodule) = self.resolve_submodule(db, name) {
-                return Place::bound(submodule).into();
-            }
+        let mut submodule_type = None;
+
+        let available_submodule_kind = self
+            .available_submodule_attributes(db)
+            .find_map(|(attr, kind)| (attr == name).then_some(kind));
+
+        if available_submodule_kind.is_some() {
+            submodule_type = self.resolve_submodule(db, name);
+        }
+
+        // if we're in a module `foo` and `foo` contains `import a.b`,
+        // and the package `a` has a submodule `b`, we assume that the
+        // attribute access `a.b` inside `foo` will resolve to the submodule
+        // `a.b` *even if* `a/__init__.py` also defines a symbol `b` (e.g. `b = 42`).
+        // This is a heuristic, but it's almost certainly what will actually happen
+        // at runtime. However, if `foo` only contains `from a.b import <something>,
+        // we prioritise the `b` attribute in `a/__init__.py` over the submodule `a.b`.
+        if available_submodule_kind == Some(ImportKind::Import)
+            && let Some(submodule) = submodule_type
+        {
+            return Place::bound(submodule).into();
         }
 
         let place_and_qualifiers = self
             .module(db)
             .file(db)
-            .map(|file| imported_symbol(db, file, name, None))
+            .map(|file| {
+                imported_symbol(db, file, name, None).map_type(|ty| {
+                    if let Some(importing) = self.importing_file(db)
+                        && let Type::ModuleLiteral(module) = ty
+                    {
+                        Type::module_literal(db, importing, module.module(db))
+                    } else {
+                        ty
+                    }
+                })
+            })
```

Comment:
it's extremely hard to write a regression test for this (though I might try harder later), but empirically this gets rid of a false-positive diagnostic on [this line](https://github.com/DataDog/dd-trace-py/blob/7e78fe5f172442f2379ea725fe4e5ea534f00395/ddtrace/contrib/internal/trace_utils.py#L56) that was ultimately due to us thinking the `importing_file` of `ddtrace.internal` was `ddtrace/__init__.py` rather than `ddtrace/contrib/internal/trace_utils.py`